prov/mrail/src/mrail_rma.c

```diff
@@ -0,0 +1,253 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "mrail.h"
+
+typedef ssize_t mrail_rma_msg_fn(struct fid_ep *ep_fid,
+		const struct fi_msg_rma *msg, uint64_t flags);
+
+static ssize_t mrail_ep_post_rma(struct fid_ep *ep_fid,
+		const struct fi_msg_rma *msg, uint64_t flags,
+		mrail_rma_msg_fn *rma_fn)
+{
+	struct mrail_ep *mrail_ep;
+	struct mrail_addr_key *mr_map;
+	fi_addr_t *rail_fi_addr;
+	uint32_t rail;
+	ssize_t ret;
+
+	size_t num_rails;
+	size_t remaining_len;
+	size_t chunk_len;
+	size_t offset;
+	size_t sublen;
+	size_t i;
+
+	struct mrail_rma_req *req;
+
+	if (msg->iov_count > 1) {
+		FI_WARN(&mrail_prov, FI_LOG_EP_DATA,
+				"Cannot handle iov_count > 1");
+		return -FI_ENOSYS;
+	}
+
+	if (msg->rma_iov_count > 1) {
+		FI_WARN(&mrail_prov, FI_LOG_EP_DATA,
+				"Cannot handle rma_iov_count > 1");
+		return -FI_ENOSYS;
+	}
+
+	mrail_ep = container_of(ep_fid, struct mrail_ep, util_ep.ep_fid.fid);
+	rail_fi_addr = ofi_av_get_addr(mrail_ep->util_ep.av, (int)msg->addr);
+	mr_map = (struct mrail_addr_key*)msg->rma_iov->key;
+
+	num_rails = mrail_ep->num_eps;
+	remaining_len = msg->msg_iov->iov_len;
+	chunk_len = remaining_len / num_rails;
+
+	req = calloc(1, sizeof(*req) + num_rails * sizeof(*req->subreqs));
+	if (!req) {
+		return -FI_ENOMEM;
+	}
+
+	ofi_atomic_initialize32(&req->remaining_subcomps, num_rails);
+	req->comp.op_context	= msg->context;
+	req->comp.flags		= flags;
+
+	offset = 0;
+	for (i=0; i<num_rails; ++i) {
+		struct iovec rail_iovec;
+		struct fi_msg_rma rail_msg;
+		struct fi_rma_iov rail_rma_iov;
+
+		sublen = remaining_len;
+		if ((remaining_len > chunk_len) && (i < (num_rails-1))) {
+			sublen = chunk_len;
+		}
+
+		req->subreqs[i].parent = req;
+
+		rail = mrail_get_rma_rail(mrail_ep);
+
+		rail_iovec.iov_base	=
+			(void*)((uintptr_t)msg->msg_iov->iov_base + offset);
+		rail_iovec.iov_len	= sublen;
+
+		rail_rma_iov.addr	= msg->rma_iov->addr + offset;
+		rail_rma_iov.len	= sublen;
+		rail_rma_iov.key	= mr_map[rail].key;
+
+		rail_msg.msg_iov	= &rail_iovec;
+		rail_msg.desc		= NULL;
+		rail_msg.iov_count 	= msg->iov_count;
+		rail_msg.addr 		= rail_fi_addr[rail];
+		rail_msg.rma_iov 	= &rail_rma_iov;
+		rail_msg.rma_iov_count 	= msg->rma_iov_count;
+		rail_msg.context 	= &req->subreqs[i].context;
+		rail_msg.data 		= msg->data;
+
+		ret = rma_fn(mrail_ep->rails[rail].ep, &rail_msg, flags);
+		if (ret) {
+			FI_WARN(&mrail_prov, FI_LOG_EP_DATA,
+				"Unable to post RMA on rail: %" PRIu32
+				" ret=%ld\n",
+				rail, ret);
+			goto error;
```

Comment:
We need to be able to handle one rail being busy and queuing the remainder of the request if needed, or place on another rail.  Alternatively, we need to ensure that all rails are available (not sure if this is feasible).